File: packages/llmix/typescript/src/config-registry-types.ts
import type { DidWebVerifier, IntegrityField, RekorClient, SignatureEntry, SigstoreVerifier, TrustPolicy } from "@snoai/mda-config"

import type { MdaConfigLoadOptions } from "./mda-loader.js"

export const MANIFEST_SCHEMA_VERSION = 1
export const REGISTRY_ROOT_SCHEMA = "llmix.config-registry.root"
export const REGISTRY_ROOT_SCHEMA_VERSION = 1
export const REGISTRY_ROOT_ENVELOPE_SCHEMA = "llmix.config-registry.root-envelope"
export const REGISTRY_ROOT_ENVELOPE_SCHEMA_VERSION = 1
export const REGISTRY_ROOT_PAYLOAD_TYPE = "application/vnd.snoai.llmix.registry-root+json"
export const REGISTRY_ROOT_FILENAME = "registry-root.json"
export const REVISION_PATTERN = /^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$/
export const REVISION_TOKEN_PATTERN = /\d+|\D+/g
export const DIGIT_TOKEN_PATTERN = /^\d+$/
export const SHA256_PATTERN = /^[a-f0-9]{64}$/

export type ConfigRegistryJsonValue = null | boolean | number | string | ConfigRegistryJsonValue[] | ConfigRegistryJsonObject

export interface ConfigRegistryJsonObject {
	[key: string]: ConfigRegistryJsonValue
}

export type JsonValue = ConfigRegistryJsonValue
export interface JsonObject extends ConfigRegistryJsonObject {}

export interface PresetSource {
	module: string
	preset: string
	presetId: string
	sourcePath: string
}

export interface ManifestPresetEntry {
	source_path: string
	source_sha256: string
	resolved_path: string
	resolved_sha256: string
}

export interface RegistryManifest {
	revision: string
	published_at: string
	schema_version: number
	presets: Record<string, ManifestPresetEntry>
}

export interface CurrentPointer {
	revision: string
	manifestSha256: string
}

export interface ParsedCurrentPointer {
	revision: string
	manifestSha256: string | null
}

export interface PublishedRevision {
	revision: string
	compiledPath: string
	manifestPath: string
	manifestSha256: string
	registryRootPath?: string
	registryRootSha256?: string
	activated: boolean
	presetIds: string[]
}

export interface RegistryRootCurrentBinding extends ConfigRegistryJsonObject {
	path: "current.json"
	revision: string
	manifest_sha256: string
	sha256: string
}

export interface RegistryRootManifestBinding extends ConfigRegistryJsonObject {
	path: string
	sha256: string
}

export interface RegistryRootFileDigest extends ConfigRegistryJsonObject {
	path: string
	sha256: string
	role: "source" | "resolved"
}

export interface RegistryRootPayload extends ConfigRegistryJsonObject {
	schema: typeof REGISTRY_ROOT_SCHEMA
	schema_version: typeof REGISTRY_ROOT_SCHEMA_VERSION
	revision: string
	published_at: string
	current: RegistryRootCurrentBinding
	manifest: RegistryRootManifestBinding
	files: RegistryRootFileDigest[]
}

export interface RegistryRootIntegrity extends IntegrityField, ConfigRegistryJsonObject {
	algorithm: "sha256"
	digest: string
}

export interface RegistryRootSignature extends SignatureEntry, ConfigRegistryJsonObject {
	signer: string
	"key-id": string
	"payload-digest": string
	algorithm: "ed25519" | "ecdsa-p256" | "rsa-pss-sha256"
	signature: string
	"rekor-log-id"?: string
	"rekor-log-index"?: number
	"payload-type": string
}

export interface RegistryRootEnvelope extends ConfigRegistryJsonObject {
	schema: typeof REGISTRY_ROOT_ENVELOPE_SCHEMA
	schema_version: typeof REGISTRY_ROOT_ENVELOPE_SCHEMA_VERSION
	payload: RegistryRootPayload
	integrity: RegistryRootIntegrity
	payload_sha256: string
	signatures: RegistryRootSignature[]
}

export interface RegistryRootSigningInput {
	payload: RegistryRootPayload
	canonicalPayload: string
	integrity: RegistryRootIntegrity
	payloadType: typeof REGISTRY_ROOT_PAYLOAD_TYPE
	payloadSha256: string
}

export type RegistryRootSigner = (
	input: RegistryRootSigningInput,
) => Promise<RegistryRootSignature | readonly RegistryRootSignature[]> | RegistryRootSignature | readonly RegistryRootSignature[]

export interface RegistryRootSigningOptions {
	signer: RegistryRootSigner
	minSignatures?: number
}

export interface RegistryRootFreshnessInput extends RegistryRootSigningInput {
	envelope: RegistryRootEnvelope
}

export type RegistryRootHighWatermark = (
	input: RegistryRootFreshnessInput,
) => Promise<boolean> | boolean

export interface RegistryRootVerificationOptions {
	trustPolicy: TrustPolicy
	rekorClient?: RekorClient
	sigstoreVerifier?: SigstoreVerifier
	didWebVerifier?: DidWebVerifier
	expectedRevision?: string
	expectedRootDigest?: string
	minimumRevision?: string
	minimumPublishedAt?: string | Date
	highWatermark?: RegistryRootHighWatermark
}

export interface ConfigRegistryOpenOptions {
	signedRoot?: RegistryRootVerificationOptions
}

export interface ConfigRegistryPublishOptions extends MdaConfigLoadOptions {
	revision?: string
	activate?: boolean
	registryRoot?: RegistryRootSigningOptions
}

File: packages/llmix/python/llmix/config_registry_types.py
"""Config Registry public and internal data types."""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
from typing import Any

REGISTRY_ROOT_SCHEMA = "llmix.config-registry.root"
REGISTRY_ROOT_SCHEMA_VERSION = 1
REGISTRY_ROOT_ENVELOPE_SCHEMA = "llmix.config-registry.root-envelope"
REGISTRY_ROOT_ENVELOPE_SCHEMA_VERSION = 1
REGISTRY_ROOT_PAYLOAD_TYPE = "application/vnd.snoai.llmix.registry-root+json"
REGISTRY_ROOT_FILENAME = "registry-root.json"


@dataclass(frozen=True)
class PublishedRevision:
    """Result of a successful registry publish."""

    revision: str
    compiled_path: Path
    manifest_path: Path
    manifest_sha256: str
    activated: bool
    preset_ids: tuple[str, ...]
    registry_root_path: Path | None = None
    registry_root_sha256: str | None = None


@dataclass(frozen=True)
class ConfigRegistryPublishOptions:
    """Options for publishing source MDA into an immutable compiled revision."""

    verify_integrity: bool = False
    verify_signatures: bool = False
    trusted_runtime: bool = False
    enforce_requires: bool = False
    allowed_networks: list[str] | None = None
    trust_policy: Any | None = None
    rekor_client: Any | None = None
    sigstore_verifier: Any | None = None
    did_web_verifier: Any | None = None
    registry_root: RegistryRootSigningOptions | None = None


@dataclass(frozen=True)
class _PresetSource:
    module: str
    preset: str
    preset_id: str
    source_path: Path


@dataclass(frozen=True)
class RegistryRootSigningInput:
    payload: dict[str, Any]
    canonical_payload: str
    integrity: dict[str, str]
    payload_type: str
    payload_sha256: str


RegistryRootSigner = Callable[
    [RegistryRootSigningInput],
    dict[str, Any] | list[dict[str, Any]] | tuple[dict[str, Any], ...],
]


@dataclass(frozen=True)
class RegistryRootSigningOptions:
    signer: RegistryRootSigner
    min_signatures: int = 1


@dataclass(frozen=True)
class RegistryRootFreshnessInput(RegistryRootSigningInput):
    envelope: dict[str, Any]


RegistryRootHighWatermark = Callable[[RegistryRootFreshnessInput], bool]


@dataclass(frozen=True)
class RegistryRootVerificationOptions:
    trust_policy: Any
    rekor_client: Any | None = None
    sigstore_verifier: Any | None = None
    did_web_verifier: Any | None = None
    expected_revision: str | None = None
    expected_root_digest: str | None = None
    minimum_revision: str | None = None
    minimum_published_at: str | datetime | None = None
    high_watermark: RegistryRootHighWatermark | None = None


@dataclass(frozen=True)
class ConfigRegistryOpenOptions:
    signed_root: RegistryRootVerificationOptions | None = None
